californium-core/src/main/java/org/eclipse/californium/core/CoapClient.java

```diff
@@ -846,7 +849,15 @@ private Request ifNoneMatch(Request request) {
 	private CoapObserveRelation observeAndWait(Request request, CoapHandler handler) {
 		Endpoint outEndpoint = getEffectiveEndpoint(request);
 		CoapObserveRelation relation = new CoapObserveRelation(request, outEndpoint);
-		request.addMessageObserver(new ObserveMessageObserverImpl(handler, relation));
+		// add message observer to get the response.
+		ObserveMessageObserverImpl messageObserver = new ObserveMessageObserverImpl(handler, relation);
+		request.addMessageObserver(messageObserver);
+		// add notification listener to all notification
+		NotificationListener notificationListener = new Adapter(messageObserver, request);
+		outEndpoint.addNotificationListener(notificationListener);
+		// messageobserver should remove this listener when the request is cancelled
+		messageObserver.setNotificationListener(notificationListener);
+		relation.setNotificationListener(notificationListener);
```

Comment:
Well, the `RedisObservationStore` could itself implement the in-memory (first level) cache. When the observation is canceled, i.e. `RedisObservationStore.remove()` is invoked, you simply remove it from the map as well as from Redis :-)
This way we do not force anybody to do anything because the in-memory cache is part of the implementation anyways. If somebody else wants to implement `ObservationStore` in a different way they are still free to do so. I am actually following a very similar approach for [implementing DTLS session caching](https://github.com/eclipse/californium/tree/session_cache) ...

My concern with the adapter is that for every observation we register a new `NotificationListener` that needs to be invoked when a notification arrives. Thus, the number of listeners grows linearly with the number of connected (and observed) devices...